sql/internet_outages/italy_covid19_outage_v1/query.sql

```diff
@@ -0,0 +1,193 @@
+DECLARE country_code STRING DEFAULT "IT";
+DECLARE fromdate DATE DEFAULT '2020-03-01'; -- '2019-12-01'
+DECLARE todate DATE DEFAULT '2020-03-02'; --'2020-03-31';
+
+-- The minimum size of a bucket for it to be published.
+DECLARE MIN_NUMBER_CLIENTS INT64 DEFAULT 5000;
+
+-- Note: udf.udf_json_extract_int_map map doesn't work in this case as it expects an INT -> INT
+-- map, while we have a STRING->int map
+CREATE TEMP FUNCTION
+  udf_json_extract_string_to_int_map (input STRING) AS (ARRAY(
+    SELECT
+      STRUCT(CAST(SPLIT(entry, ':')[OFFSET(0)] AS STRING) AS key,
+             CAST(SPLIT(entry, ':')[OFFSET(1)] AS INT64) AS value)
+    FROM
+      UNNEST(SPLIT(REPLACE(TRIM(input, '{}'), '"', ''), ',')) AS entry
+    WHERE
+      LENGTH(entry) > 0 ));
+     
+
+-- Get a stable source for DAUs.
+WITH DAUs AS (
+    SELECT
+        submission_date as date,
+        COUNT(*) as client_count
+    FROM
+        telemetry.clients_daily
+    WHERE
+        submission_date >= fromdate
+        AND submission_date <= todate
+        AND country = country_code
+    GROUP BY 1
+),
+-- Compute aggregates for the health data.
+health_data_sample AS (
+    SELECT 
+        DATE(SAFE_CAST(creation_date as TIMESTAMP)) as date,
+        client_id,
+        SUM(coalesce(CAST(JSON_EXTRACT(additional_properties, '$.payload.sendFailure.undefined') AS INT64),0)) as eUndefined,
+        SUM(coalesce(CAST(JSON_EXTRACT(additional_properties, '$.payload.sendFailure.timeout') AS INT64),0)) as eTimeOut,
+        SUM(coalesce(CAST(JSON_EXTRACT(additional_properties, '$.payload.sendFailure.abort') AS INT64),0)) as eAbort,
+        SUM(coalesce(CAST(JSON_EXTRACT(additional_properties, '$.payload.sendFailure.eUnreachable') AS INT64),0)) as eUnreachable,
+        SUM(coalesce(CAST(JSON_EXTRACT(additional_properties, '$.payload.sendFailure.eTerminated') AS INT64),0))  as eTerminated,
+        SUM(coalesce(CAST(JSON_EXTRACT(additional_properties, '$.payload.sendFailure.eChannelOpen') AS INT64),0))  as eChannelOpen,
+    FROM
+        telemetry.health
+    WHERE
+        date(submission_timestamp) >= fromdate
+        AND date(submission_timestamp) <= todate
+        AND normalized_country_code = country_code
+    GROUP BY 1, 2
+    --LIMIT 10 -- test only: REMOVE ME
+), health_data_aggregates AS (
+    SELECT
+        date,
+        COUNTIF(eUndefined > 0) as num_clients_eUndefined,
+        COUNTIF(eTimeOut > 0) as num_clients_eTimeOut,
+        COUNTIF(eAbort > 0) as num_clients_eAbort,
+        COUNTIF(eUnreachable > 0) as num_clients_eUnreachable,
+        COUNTIF(eTerminated > 0) as num_clients_eTerminated,
+        COUNTIF(eChannelOpen > 0) as num_clients_eChannelOpen,
+    FROM health_data_sample
+    GROUP BY date
+    HAVING
+      COUNT(*) > MIN_NUMBER_CLIENTS
+), final_health_data AS (
+    SELECT
+        h.date,
+        (num_clients_eUndefined / DAUs.client_count) as proportion_undefined,
```

Comment:
Another option to make aggregates even safer is to round any ratio < 0.01 (1% of the reference user-base) to 0.

Reply:
This didn't turn out to be required at data-review time.